mp2-v1/tests/common/cases/query.rs

```diff
@@ -896,41 +895,61 @@ pub async fn prove_non_existence_row<'a>(
             .table
             .execute_row_query(&query.unwrap(), &[])
             .await?;
+
         if rows.len() == 0 {
             // no node found, return None
             info!("Search node for non-existence circuit: no node found");
             return Ok(None);
         }
+
         let row_key = rows[0]
             .get::<_, Option<Vec<u8>>>(0)
             .map(RowTreeKey::from_bytea)
             .context("unable to parse row key tree")
             .expect("");
+
         // among the nodes with the same index value of the node with `row_key`, we need to find
         // the one in the highest place in the tree, i.e., a node whose parent has not the same
         // index value
-        let (mut node_ctx, node_value) = row_tree
-            .fetch_with_context_at(&row_key, primary as Epoch)
-            .await;
-        let value = node_value.value();
-        let get_parent_data = async |node_ctx: &NodeContext<RowTreeKey>| {
-            if node_ctx.parent.is_some() {
-                let parent_key = node_ctx.parent.as_ref().unwrap();
-                let (ctx, value) = row_tree
-                    .fetch_with_context_at(parent_key, primary as Epoch)
-                    .await;
-                Some((ctx, value))
+        let primary_epoch = primary as Epoch;
+        let lineage = row_tree
+            .tree()
+            .lineage(&row_key, &row_tree.at(primary_epoch))
+            .await
+            .unwrap();
+
+        // fetch all the lineage node payloads in a single query
+        let all_payloads = row_tree
+            .try_fetch_many_at(lineage.full_path().cloned().map(|k| (primary_epoch, k)))
+            .await
+            .with_context(|| format!("while fetching ascendance payloads for `{:?}`", row_key))?
+            .into_iter()
+            // accumulate everything in a single key -> payload hashmap, as we
+            // know for sure everything is in primary_epoch.
+            .filter_map(|maybe_data| {
+                maybe_data.map(|(epoch, k, v)| {
+                    // safety check, we should never encounter another epoch
+                    assert_eq!(epoch, primary_epoch);
+                    (k, v)
+                })
+            })
+            .collect::<HashMap<_, _>>();
```

Comment:
Could we put that in ryhope/parsil directly ? I don't see what we could do else with the row results.